x/denommetadata/ibc_middleware.go

```diff
@@ -0,0 +1,111 @@
+package denommetadata
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+
+	"github.com/dymensionxyz/dymension-rdk/x/denommetadata/types"
+)
+
+var _ porttypes.IBCModule = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the transfer middleware
+type IBCMiddleware struct {
+	porttypes.IBCModule
+	bankKeeper     types.BankKeeper
+	transferKeeper types.TransferKeeper
+	hooks          types.MultiDenomMetadataHooks
+}
+
+// NewIBCMiddleware creates a new IBCMiddleware given the keeper and underlying application
+func NewIBCMiddleware(
+	app porttypes.IBCModule,
+	bankKeeper types.BankKeeper,
+	transferKeeper types.TransferKeeper,
+	hooks types.MultiDenomMetadataHooks,
+) IBCMiddleware {
+	return IBCMiddleware{
+		IBCModule:      app,
+		bankKeeper:     bankKeeper,
+		transferKeeper: transferKeeper,
+		hooks:          hooks,
+	}
+}
+
+// OnRecvPacket registers the denom metadata if it does not exist.
+// It will intercept an incoming packet and check if the denom metadata exists.
+// If it does not, it will register the denom metadata.
+// The handler will expect a 'transferinject' object in the memo field of the packet.
+// If the memo is not an object, or does not contain the metadata, it moves on to the next handler.
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	packetData := new(transfertypes.FungibleTokenPacketData)
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), packetData); err != nil {
+		err = errorsmod.Wrapf(errortypes.ErrJSONUnmarshal, "unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	denomTrace := transfertypes.ParseDenomTrace(packetData.Denom)
+	// if denom trace path is empty (sending chain's native coin, e.g. 'adym'),
```

Comment:
wdym here? why would u get "adym" on the RA without the IBC path?
maybe add some example or test case even better